test/libsolidity/semanticTests/builtins/balance_other_contract.sol

```diff
@@ -0,0 +1,23 @@
+contract Other {
+    constructor() payable {
+    }
+    function getAddress() public returns (address) {
+        return address(this);
+    }
+}
+contract ClientReceipt {
+    Other other;
+    constructor() payable {
+        other = new Other{value:500}();
+    }
+    function getAddress() public returns (address) {
+        return other.getAddress();
+    }
+}
+// ====
+// compileViaYul: also
+// ----
+// constructor(), 2000 wei ->
+// contract_balance -> 1500
+// getAddress() -> 0xf01f7809444bd9a93a854361c6fae3f23d9e23db
```

Comment:
Is this always guaranteed? When could it change?

Reply:
Yes, thats guaranteed. The address is deterministically calculated by the address of the creator and its nonce. These things are defined in `ExecutionFramework`. The sender address is always the same (it uses `h160 ExecutionFramework::account(size_t _idx)` with an `_idx` value of `0`. `EvmHost::reset()` is calling `accounts.clear()`, that is a `std::unordered_map<address, MockedAccount>`, where `MockedAccount` defines the `nonce`. The nonce is incremented with each creation call. See `EVMHost::call(evmc_message const& _message)`. 

```
		h160 createAddress(keccak256(
			bytes(begin(message.sender.bytes), end(message.sender.bytes)) +
			asBytes(to_string(sender.nonce++))
		));
```